src/core/ext/transport/chttp2/client/secure/secure_channel_create.c

```diff
@@ -60,69 +60,57 @@ static void client_channel_factory_unref(
 static grpc_subchannel_args *get_secure_naming_subchannel_args(
     grpc_exec_ctx *exec_ctx, const grpc_subchannel_args *args) {
   grpc_channel_credentials *channel_credentials =
-      grpc_find_channel_credentials_in_args(args->args);
-  if (channel_credentials == NULL) return NULL;
-
+      grpc_channel_credentials_find_in_args(args->args);
+  // Make sure security connector does not already exist in args.
+  if (grpc_security_connector_find_in_args(args->args) != NULL) {
+    gpr_log(GPR_ERROR, "Security connector already present in channel args.");
+    return NULL;
+  }
+  if (channel_credentials == NULL) {
+    gpr_log(GPR_ERROR,
+            "Can't create subchannel: channel credentials missing for secure "
+            "channel.");
+    return NULL;
+  }
   const grpc_slice_hash_table *targets_info =
       grpc_glb_targets_info_find_in_args(args->args);
-  if (targets_info == NULL) return NULL;
-
-  /* To which address are we connecting? */
+  // To which address are we connecting?
   const char *target_uri_str = grpc_get_subchannel_address_uri_arg(args->args);
   grpc_uri *target_uri =
       grpc_uri_parse(target_uri_str, true /* suppress errors */);
-  /* Find the balancer name for the target. */
-  const grpc_slice key = grpc_slice_from_static_string(target_uri->path);
-  const char *secure_name = grpc_slice_hash_table_get(targets_info, key);
-  grpc_slice_unref_internal(exec_ctx, key);
+  char *secure_name = NULL;
+  if (targets_info != NULL) {  // LB channel
+    // Find the balancer name for the target.
+    const grpc_slice key = grpc_slice_from_static_string(target_uri->path);
+    secure_name = gpr_strdup(grpc_slice_hash_table_get(targets_info, key));
+    grpc_slice_unref_internal(exec_ctx, key);
+  } else {  // regular channel: the secure name is the address itself.
+    secure_name = gpr_strdup(target_uri->path);
```

Comment:
I think this is incorrect.  We should use the server name here, not the IP address of the individual subchannel.  It needs to come from GRPC_ARG_SERVER_URI.